### PR TITLE
Automatically emits metrics with log counts by level

### PR DESCRIPTION
**Pull request by glibsm**
The idea here is that we'll be getting a strong signal post deploy
if the distribution of `error` logs shoots up - probably revert is 
on the cards.

GFM-154

### CODE REVIEW

File: service/service_setup.go [hunk 1]
```diff
@@ -33,22 +33,23 @@ import (
 	"github.com/pkg/errors"
 )
 
-func (svc *serviceCore) setupLogging(cfg config.Provider) {
+func (svc *serviceCore) setupLogging() {
 	if svc.log == nil {
```
Comment (sectioneight) [inline service/service_setup.go]: semi-related nit: prefer early returns and reducing nesting:

```go
if svc.log != nil {
  svc.log.Debug("Using custom...")
  return
}

// the rest
logBuilder := ...
```

File: service/service_setup.go [hunk 2]
```diff
@@ -33,22 +33,23 @@ import (
...
+
 		svc.log = logBuilder.WithConfiguration(svc.logConfig).Build()
 	} else {
 		svc.log.Debug("Using custom log provider due to service.WithLogger option")
 	}
 }
 
-func (svc *serviceCore) setupStandardConfig(cfg config.Provider) error {
-	if err := cfg.Get(config.Root).PopulateStruct(&svc.standardConfig); err != nil {
+func (svc *serviceCore) setupStandardConfig() error {
```
Comment (sectioneight) [inline service/service_setup.go]: why the change to rely on the member field instead of arguments? personal preference?

File: testutils/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
```
Comment (sectioneight) [inline testutils/metrics/metrics.go]: 2017

File: testutils/metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	"time"
+
+	"github.com/uber-go/tally"
+)
+
+// TestStatsReporter allows for easy testing of any metrics related functionality
+type TestStatsReporter struct {
+	m sync.Mutex
+
+	Cw sync.WaitGroup
```
Comment (sectioneight) [inline testutils/metrics/metrics.go]: I don't like thesenames. They convey absolutely zero meaning. Additionally the casing implies it's an abbreviated word. I'd suggest `CountersWG` or similar

File: service/service_setup.go [hunk 5]
```diff
@@ -33,22 +33,23 @@ import (
 	"github.com/pkg/errors"
 )
 
-func (svc *serviceCore) setupLogging(cfg config.Provider) {
+func (svc *serviceCore) setupLogging() {
 	if svc.log == nil {
-		logBuilder := ulog.Builder()
-		// load and configure logging
+		logBuilder := ulog.Builder().WithScope(svc.metrics)
```
Comment (sectioneight) [inline service/service_setup.go]: This needs to be configurable. It should be on by default, but there needs to be a boolean flag somebody can set that turns off the metrics reporting for log messages.

File: ulog/builder.go [hunk 6]
```diff
@@ -87,6 +96,7 @@ func (lb *LogBuilder) WithSentryHook(hook *sentry.Hook) *LogBuilder {
 }
 
 // Build the ulog logger for use
+// TODO: build should return `(Log, error)` in case we can't properly instantiate
```
Comment (sectioneight) [inline ulog/builder.go]: Why a TODO and not part of this PR? What are the cases where we can't properly instantiate?

File: ulog/builder_test.go [hunk 7]
```diff
@@ -147,6 +148,22 @@ func TestConfiguredLoggerWithSentryError(t *testing.T) {
...
+	defer env.Override(t, config.EnvironmentKey(), "wat?")()
+
+	s, r := metrics.NewTestScope()
+	l := Builder().WithScope(s).Build()
+
+	r.Cw.Add(1)
+	l.Warn("Warning log!")
+	r.Cw.Wait()
+
+	assert.Equal(t, 1, len(r.Counters))
```
Comment (sectioneight) [inline ulog/builder_test.go]: Can we check that the counter used the right bucket?

File: ulog/metrics/metrics.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"github.com/uber-go/tally"
+	"github.com/uber-go/zap"
+)
+
+// Hook counts the number of logging messages per level
+func Hook(s tally.Scope) zap.Hook {
+	return zap.Hook(func(e *zap.Entry) error {
+		// TODO: consider not using .Counter method which does a map lookup
+		// and just create objects for a counter of each type and retain in a struct
```
Comment (sectioneight) [inline ulog/metrics/metrics.go]: huh?

File: service/service.go [hunk 9]
```diff
@@ -114,30 +114,30 @@ func New(options ...Option) (Owner, error) {
 		svc.configProvider = config.Load()
 	}
 
-	svc.setupLogging(svc.configProvider)
+	// Initialize metrics. If no metrics reporters were Registered, do nop
+	// TODO(glib): add a logging reporter and use it by default, rather than nop
+	svc.setupMetrics()
```
Comment (alsamylkin) [inline service/service.go]: Shouldn't we setup standard config first(it has service name, owner, all the important things), all metrics and then logger?

File: service/service_setup.go [hunk 10]
```diff
@@ -66,13 +67,13 @@ func (svc *serviceCore) setupMetrics() {
...
 
-func (svc *serviceCore) setupRuntimeMetricsCollector(cfg config.Provider) error {
+func (svc *serviceCore) setupRuntimeMetricsCollector() error {
 	if svc.RuntimeMetricsCollector() != nil {
 		return nil
 	}
 
 	var runtimeMetricsConfig metrics.RuntimeConfig
-	err := cfg.Get("metrics.runtime").PopulateStruct(&runtimeMetricsConfig)
+	err := svc.configProvider.Get("metrics.runtime").PopulateStruct(&runtimeMetricsConfig)
```
Comment (alsamylkin) [inline service/service_setup.go]: I think it should be .Scope("metrics").Get("runtime").

File: ulog/metrics/metrics.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+package metrics
+
+import (
+	"github.com/uber-go/tally"
+	"github.com/uber-go/zap"
+)
+
+// Hook counts the number of logging messages per level
+func Hook(s tally.Scope) zap.Hook {
+	return zap.Hook(func(e *zap.Entry) error {
```
Comment (alsamylkin) [inline ulog/metrics/metrics.go]: check e != nil?

File: service/service_setup.go (hunk 5 above)
Comment (glibsm) [inline service/service_setup.go]: I agree. I did spend some time thinking about it, but I am not sure with the current state of things the implementation is going to be the cleanest. 

Right now, the thing that would fit best is to include a `disabled:` flag in config and allow user to specify that. However, I don't think that mixing and matching `enabled` and `disabled` in configuration is a good thing (I've found that the positive bools tend to work out better in the long run).

I think that a better fit for this is going to be the default configuration, which is currently in the works by @alsamylkin. Then, we can include `metrics: logging: enabled: true` or something like that, and allow to easily override in service config.

Basically it's a long way of saying I think we should wait to add that in the proper way. Thoughts?

File: service/service_setup.go (hunk 2 above)
Comment (glibsm) [inline service/service_setup.go]: Mostly simplifying for three reasons:

1. the parameters being passed in are totally redundant on a method call.
2. most of these functions were not even using the parameters and used the object directly
3. needed access to the metrics scope in the logging function and rather than adding an additional parameter, simplified the whole approach.

I would be ok to restoring parameter passing, but in that case I think these methods should be changed to function and take service as a parameter. Kind of like the options pattern

File: testutils/metrics/metrics.go (hunk 3 above)
Comment (glibsm) [inline testutils/metrics/metrics.go]: need to update the license binary i guess

File: testutils/metrics/metrics.go (hunk 4 above)
Comment (glibsm) [inline testutils/metrics/metrics.go]: Point taken, i don't like them either. Will clean up

File: ulog/builder_test.go (hunk 7 above)
Comment (glibsm) [inline ulog/builder_test.go]: That is tested elsewhere, I think it would be redundant. The test here is to make sure that we set up the hook at all

File: ulog/metrics/metrics.go (hunk 8 above)
Comment (glibsm) [inline ulog/metrics/metrics.go]: engrish.

I did a horrid job of saying that we should essentially cache the counters, since there is a finite number of them: one per each level, rather than calling `Counter()` method on the scope, which does some locking and map lookup

File: testutils/metrics/metrics.go [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+import (
+	"sync"
+	"time"
+
+	"github.com/uber-go/tally"
+)
+
+// TestStatsReporter allows for easy testing of any metrics related functionality
+type TestStatsReporter struct {
```
Comment (glibsm) [inline testutils/metrics/metrics.go]: It's used by two different packages in uberfx

File: service/service.go (hunk 9 above)
Comment (glibsm) [inline service/service.go]: Yes, that's a great point.

**Comment by glibsm** [comment]
Fixing the license added a bit of fluff to the diff, can move out into another if there are objections.

File: ulog/builder.go (hunk 6 above)
Comment (glibsm) [inline ulog/builder.go]: Didn't want to bloat the PR and touch that code, but will at some point. We do some unmarhsaling, we read configs and do some other things that produce errors in logger build. When encountering error, we log an error message and fall back to default value. I think we should probably just return those errors to make sure service owners don't misconfigure logging by accident

File: ulog/builder_test.go [hunk 13]
```diff
@@ -147,6 +148,49 @@ func TestConfiguredLoggerWithSentryError(t *testing.T) {
...
+		{
+			true,
+			[]byte("logging:\n  disableMetrics: true"),
+		},
+	}
+
+	for _, tt := range tests {
+		logCfg := Configuration{}
+		logYamlCfg := config.NewYAMLProviderFromBytes(tt.data)
+		logYamlCfg.Get("logging").PopulateStruct(&logCfg)
```
Comment (alsamylkin) [inline ulog/builder_test.go]: Could you test actual behavior, not a serialization?

File: ulog/builder_test.go [hunk 14]
```diff
@@ -147,6 +148,49 @@ func TestConfiguredLoggerWithSentryError(t *testing.T) {
...
+			false,
+			[]byte("logging:\n  disableMetrics: false"),
+		},
+		{
+			true,
+			[]byte("logging:\n  disableMetrics: true"),
+		},
+	}
+
+	for _, tt := range tests {
```
Comment (alsamylkin) [inline ulog/builder_test.go]: use subtests?

File: ulog/builder_test.go (hunk 13 above)
Comment (glibsm) [inline ulog/builder_test.go]: i'll remove the test because you're right it doesn't help. But testing behavior right now is not easy - since zap doesn't let you peek into the hooks or options that were used to create it. I'll separate the `Build()` method into two parts, so we can assert which options zap is about to be created with.
